### ISSUE TITLE
Introduce Arbitrary Delays for Topics

### ISSUE THREAD

**Issue by tomazfernandes** [spring-projects/spring-kafka#2247]
Now that we have the `KafkaConsumerBackOffManager`, we can build an arbitrary delay mechanism for topics. This is a common functionality that inter-service communication systems usually provide and that AFAIK [has been missing](https://support.huaweicloud.com/intl/en-us/kafka_faq/kafka-faq-0018.html) from the `Kafka` ecosystem.

There are two kinds of such mechanisms. In the first one, the `consumer` will look at the record creation timestamp and apply a predetermined delay to that record by backing its partition off if necessary until its due time.

In the second kind, the `producer` sets a `dueTimestamp` header to the record and consumers will refrain from consuming the message until then (which is what we currently use for retryable topics).

Each mechanism caters to different messaging needs.

Note that this is completely apart from the retryable topics feature - although a natural follow up would be refactoring the retryable topics to use this mechanism under the hood, and even offer the `consumer` approach.

Without further ado, I've been sitting on this for quite a while now so not to mix with the configuration issues, so I'll open a draft PR so you can take a look when you can. No hurry of course.

I hope you enjoy it! Thanks!

**Comment by tomazfernandes** [comment]
As a note, I'm not married to this annotation - I looked into simply adding a `delay` property to `@KafkaListener`, but then there would be this `producer` / `consumer` part, and the delay precision configuration, producer timestamp header, and I figured it'd become bloated.

Might work as a `@KafkaListener(delay = @DelayedTopic(...))` though, if that's better.

I also looked into making it configurable through the `factory`, and although it might be a nice to have, I think this should be a per-topic configuration.

I also looked into having separate annotations for `producer` and `consumer`, but it seemed it would only complicate things.

So, if you have any different ideas you like better than the annotation, or anything else you want to discuss regarding the concept as a whole, please let me know.

Thanks!

**Comment by artembilan** [comment]
Just for general info. The RabbitMQ has a Delayed Exchange feature: https://blog.rabbitmq.com/posts/2015/04/scheduling-messages-with-rabbitmq.
That one is a broker feature and our producer should just said how long to delay.
After a delay (an expiration in that exchange, essentially) the message is routed to the target exchange for consumption.

Not sure though how this is going to be correlated with Kafka partition order...
Also: https://stackoverflow.com/questions/69245388/how-to-implement-time-to-live-ttl-for-a-kafka-topic

**Comment by tomazfernandes** [comment]
> Just for general info. The RabbitMQ has a Delayed Exchange feature: https://blog.rabbitmq.com/posts/2015/04/scheduling-messages-with-rabbitmq. That one is a broker feature and our producer should just said how long to delay. After a delay (an expiration in that exchange, essentially) the message is routed to the target exchange for consumption.

Yes, thanks, I had looked into that. AWS also offers delays for SQS, although it has a queue based proper rather than a producer header delay: https://docs.aws.amazon.com/AWSSimpleQueueService/latest/SQSDeveloperGuide/sqs-delay-queues.html

So, the `PRODUCER` mode in our solution should mimic the `RabbitMQ` approach, while the `CONSUMER` mode should mimic the `SQS` approach - I think both are useful for different scenarios.

> Not sure though how this is going to be correlated with Kafka partition order...

The backing off feature relies on having a fixed delay for a given partition. Let's say it's a 5 second delay - if we consume a message that's not yet due, we can be sure all other messages in that partition are also not due, since they came in later. So after the given time we consume that message and look up the due time for the next one, and so on.

We can actually only guarantee that the message will be consumed after the delay, not at the exact time, otherwise we'd need to process records in parallel and thus lose ordering guarantees.

> Also: https://stackoverflow.com/questions/69245388/how-to-implement-time-to-live-ttl-for-a-kafka-topic

I'm not sure I understand what in this question is related to this issue - it seems to be target at topic deletion rather than message delaying.

**Comment by artembilan** [comment]
Yeah... Nevermind for TTL. 
You have explained it perfectly: if message is delayed, we just pause consumer and back off it for the provided delay.
Sounds reasonable.

I'm not sure in extra annotation. Perhaps just single property on the `@KafkaListener` like `delayStrategy`. Just because there might be cases when neither consumer, nor producer are appropriate for the target solution.
This way that enum fully does not make sense for end-user. And so on, so on...
I will need to look into this closer for more thoughts.

Thank you so far!

**Comment by tomazfernandes** [comment]
Just a couple more notes on this.

We might be able to come up with `non-blocking arbitrary scheduling` of records by using a `my-topic-delay` topic with a given number of partitions. In principle, if we have 100 partitions for that topic, we'd be able to have 100 different scheduled messages at any given time for that topic, also depending of course on the number of consumers and the time a record takes to be processed.

We'd have logic to keep track of each partition's state - probably what we already have in the `BackOffManager` would be enough or at least a start, and then figure out the proper partition for which to send the record to. Maybe we'd also need to keep track of when processing of records finishes or some other way to know that a partition is or will be `free` to send a record there. Shouldn't be too hard on a `best effort` basis.

This might also be the mechanism behind what I proposed here: https://github.com/spring-projects/spring-kafka/issues/2150#issuecomment-1115189641

Not so simple solutions with lots of work, although definitely doable, so something to keep in mind that I wanted to share with you.

**Comment by tomazfernandes** [comment]
The other note: as companies increasingly migrate to `Kafka` not only for high-throughput streaming but also for general-purpose inter service communication, I expect these patterns to be in high demand, since other messaging tools provide them OOTB and there are messaging architectural patterns that rely on those. Also, they're anything but trivial to be properly implemented with `Kafka` as we well know.

That's one of the reasons I went with the `@DelayedTopic` annotation - might provide more room for growth if users requests more features to be added to it.

Ok, I'll let you focus on anything else, sorry for the noise 😄

Thanks.

**Comment by selalerercapitolis** [comment]
@tomazfernandes
I don't know of any valid scenario where such a feature is desirable. 

The only valid scenario that I know, in other technologies, is task scheduling, but that is only reasonable where you can set the schedule differently for individual tasks. With Kafka's partition order that is not possible (you can only set the same delay for the entire partition).

This feature is usually abused in other technologies to do synchronize-by-sleep which is a horrible synchronization method.

**Comment by tomazfernandes** [comment]
Hi @selalerercapitolis, thanks for looking into this. There are actually quite a few use cases for delaying messages.

In distributed systems, due to eventual consistency, sometimes you need to give another system a head start for processing something that your system will require. Since you know that trying as soon as the message arrives would likely result in an error due to the information not being available yet, it makes sense to add a delay to it to avoid wasting resources. If more attempts are needed after that, the delayed topic can be used in conjunction with `@RetryableTopic` to handle further attempts.

As you mentioned, this can also be used for scheduling tasks - you may have a requirement of starting a process let's say 12 hours after receiving an event, such as sending an email or whatnot. As long as the delay is constant, Kafka's ordering guarantees do not get in the way. I've explained above a way we could handle arbitrary scheduling in `Kafka`, but that shouldn't be too easy to implement.

Another important use-case is, as `Kafka` is being increasingly adopted for inter-service communications, companies migrating to `Kafka` often need it to behave like a traditional messaging system so they don't have to change all their systems' architectures at once.

That's what came to mind now - I've actually had to use those patterns before. I'm sure users have other real-world needs that can be addressed with this. I think it's the framework's place to provide the tools - users can make good or bad use of it, as with anything else.

Thanks.

**Comment by abhaym0109** [comment]
> As a note, I'm not married to this annotation - I looked into simply adding a `delay` property to `@KafkaListener`, but then there would be this `producer` / `consumer` part, and the delay precision configuration, producer timestamp header, and I figured it'd become bloated.
> 
> Might work as a `@KafkaListener(delay = @DelayedTopic(...))` though, if that's better.
> 
> I also looked into making it configurable through the `factory`, and although it might be a nice to have, I think this should be a per-topic configuration.
> 
> I also looked into having separate annotations for `producer` and `consumer`, but it seemed it would only complicate things.
> 
> So, if you have any different ideas you like better than the annotation, or anything else you want to discuss regarding the concept as a whole, please let me know.
> 
> Thanks!



> As a note, I'm not married to this annotation - I looked into simply adding a `delay` property to `@KafkaListener`, but then there would be this `producer` / `consumer` part, and the delay precision configuration, producer timestamp header, and I figured it'd become bloated.
> 
> Might work as a `@KafkaListener(delay = @DelayedTopic(...))` though, if that's better.
> 
> I also looked into making it configurable through the `factory`, and although it might be a nice to have, I think this should be a per-topic configuration.
> 
> I also looked into having separate annotations for `producer` and `consumer`, but it seemed it would only complicate things.
> 
> So, if you have any different ideas you like better than the annotation, or anything else you want to discuss regarding the concept as a whole, please let me know.
> 
> Thanks!

Hi, @tomazfernandes I am trying to add delay in my Kafka, that is whenever a message is to my consumer I want the consumer to wait for ('n' minutes) and then proceed wit the message. How can I achieve it?

**Comment by garyrussell** [comment]
@abhaym0109 This feature is not currently scheduled to be implemented. At best it might be considered for 3.1 next year.

If you always want to delay by the same time for each record, you can use the existing `@RetryableTopic` mechanism and throw an exception when the record is received from the main topic; it will then be written to the first retry topic and redelivered after the delay.

**Comment by tomazfernandes** [comment]
Hi @abhaym0109, as @garyrussell pointed out, it seems this shouldn't be getting support from the framework in the near future.

But there's a simple implementation you can do for now that achieves the result you're looking for.

You can implement this class:

#### DelayedMessageListenerAdapter implementation

```java

/**
 * An {@link AcknowledgingConsumerAwareMessageListener} implementation that delays
 * record consumption by backing off the record's partition until a given delay
 * past the record's creation timestamp.
 *
 * @param <K> the key.
 * @param <V> the value.
 *
 * @author Tomaz Fernandes
 *
 * @see KafkaConsumerBackoffManager
 */
public class DelayedMessageListenerAdapter<K, V>
		extends AbstractDelegatingMessageListenerAdapter<MessageListener<K, V>>
		implements AcknowledgingConsumerAwareMessageListener<K, V> {

	private final LogAccessor logger = new LogAccessor(LogFactory.getLog(getClass()));

	private static final Duration DEFAULT_DELAY_VALUE = Duration.ZERO;

	private final String listenerId;

	private final KafkaConsumerBackoffManager kafkaConsumerBackoffManager;

	private final Map<String, Duration> delaysPerTopic = new ConcurrentHashMap<>();

	private Duration defaultDelay = DEFAULT_DELAY_VALUE;

	public DelayedMessageListenerAdapter(MessageListener<K, V> delegate,
	                                      KafkaConsumerBackoffManager kafkaConsumerBackoffManager,
	                                      String listenerId) {

		super(delegate);
		Assert.notNull(kafkaConsumerBackoffManager, "kafkaConsumerBackoffManager cannot be null");
		Assert.notNull(listenerId, "listenerId cannot be null");
		this.kafkaConsumerBackoffManager = kafkaConsumerBackoffManager;
		this.listenerId = listenerId;
	}

	@Override
	public void onMessage(ConsumerRecord<K, V> consumerRecord, @Nullable Acknowledgment acknowledgment,
						  @Nullable Consumer<?, ?> consumer) throws KafkaBackoffException {

		this.kafkaConsumerBackoffManager
				.backOffIfNecessary(createContext(consumerRecord,
						consumerRecord.timestamp() + delaysPerTopic.getOrDefault(consumerRecord.topic(), this.defaultDelay).toMillis(),
						consumer));
		invokeDelegateOnMessage(consumerRecord, acknowledgment, consumer);
	}

	/**
	 * Sets the amount of time the record will be delayed for this topic,
	 * counting from its creation timestamp.
	 * <p>
	 * The delay value can be changed at runtime - mind that
	 * if consumption is already backed off, if the new value is greater than the
	 * previous value, consumption will be backed off until the new value.
	 * <p>
	 * However, if the new value is less than the current value,
	 * the new value will be effective beginning in the next record.
	 *
	 * @param topic the delay value.
	 * @param delay the delay value.
	 */
	public void setDelayForTopic(String topic, Duration delay) {
		Assert.notNull(topic, "Topic cannot be null");
		Assert.notNull(delay, "Delay cannot be null");
		this.logger.debug(() -> String.format("Setting delay %s for listener id %s", delay, this.listenerId));
		this.delaysPerTopic.put(topic, delay);
	}

	/**
	 * Sets the amount of time the record will be delayed for this adapter,
	 * counting from its creation timestamp.
	 *
	 * @param delay the delay value
	 */
	public void setDefaultDelay(Duration delay) {
		Assert.notNull(delay, "Delay cannot be null");
		this.logger.debug(() -> String.format("Setting delay %s for listener id %s", delay, this.listenerId));
		this.defaultDelay = delay;
	}

	private void invokeDelegateOnMessage(ConsumerRecord<K, V> consumerRecord, Acknowledgment acknowledgment, Consumer<?, ?> consumer) {
		switch (this.delegateType) {
			case ACKNOWLEDGING_CONSUMER_AWARE:
				this.delegate.onMessage(consumerRecord, acknowledgment, consumer);
				break;
			case ACKNOWLEDGING:
				this.delegate.onMessage(consumerRecord, acknowledgment);
				break;
			case CONSUMER_AWARE:
				this.delegate.onMessage(consumerRecord, consumer);
				break;
			case SIMPLE:
				this.delegate.onMessage(consumerRecord);
		}
	}

	private KafkaConsumerBackoffManager.Context createContext(ConsumerRecord<K, V> data, long nextExecutionTimestamp,
															  Consumer<?, ?> consumer) {

		return this.kafkaConsumerBackoffManager.createContext(nextExecutionTimestamp, this.listenerId,
				new TopicPartition(data.topic(), data.partition()), consumer);
	}

	/*
	 * Since the container uses the delegate's type to determine which method to call, we
	 * must implement them all.
	 */
	@Override
	public void onMessage(ConsumerRecord<K, V> data) {
		onMessage(data, null, null); // NOSONAR
	}

	@Override
	public void onMessage(ConsumerRecord<K, V> data, Acknowledgment acknowledgment) {
		onMessage(data, acknowledgment, null); // NOSONAR
	}

	@Override
	public void onMessage(ConsumerRecord<K, V> data, Consumer<?, ?> consumer) {
		onMessage(data, null, consumer);
	}
}
```

#### ContainerFactory configuration

```java

@Bean
public ConcurrentKafkaListenerContainerFactory<Object, Object> kafkaListenerContainerFactory(ConsumerFactory<Object, Object> consumerFactory, ListenerContainerRegistry registry, TaskScheduler scheduler) {
	ConcurrentKafkaListenerContainerFactory<Object, Object> factory = new ConcurrentKafkaListenerContainerFactory<>();
	factory.setConsumerFactory(consumerFactory);
	KafkaConsumerBackoffManager backOffManager = createBackOffManager(registry, scheduler);
	factory.setContainerCustomizer(container -> {
		DelayedMessageListenerAdapter<Object, Object> delayedAdapter = wrapWithDelayedMessageListenerAdapter(backOffManager, container);
		delayedAdapter.setDelayForTopic(TEST_TOPIC, Duration.ofSeconds(6));
		delayedAdapter.setDefaultDelay(Duration.ofSeconds(1));
		container.setupMessageListener(delayedAdapter);
	});
	return factory;
}

@Bean
public TaskScheduler taskScheduler() {
	return new ThreadPoolTaskScheduler();
}

@SuppressWarnings("unchecked")
private DelayedMessageListenerAdapter<Object, Object> wrapWithDelayedMessageListenerAdapter(KafkaConsumerBackoffManager backOffManager, ConcurrentMessageListenerContainer<Object, Object> container) {
	return new DelayedMessageListenerAdapter<>((MessageListener<Object, Object>) container.getContainerProperties().getMessageListener(), backOffManager, container.getListenerId());
}

private ContainerPartitionPausingBackOffManager createBackOffManager(ListenerContainerRegistry registry, TaskScheduler scheduler) {
	return new ContainerPartitionPausingBackOffManager(registry, new ContainerPausingBackOffHandler(new ListenerContainerPauseService(registry, scheduler)));
}
```

You can configure a delay for all topics for this factory, or add specific delays for specific topics. Default delay is zero.

Also, you can have a separate `ListenerContainerFactory` to handle delayed topics, or different delay configurations.

#### Sample Spring Boot App

Here's a sample `Spring Boot` app that delays all messages by one second and messages sent to `test.topic` by 6 seconds:

```java

@SpringBootApplication
public class DelayedTopicApp {

    private static final String TEST_TOPIC = "test.topic";

	private static final String TEST_MESSAGE = "My message";

	@Autowired
	KafkaTemplate<String, String> kafkaTemplate;

	public static void main(String[] args) {
		SpringApplication.run(DelayedTopicApp.class, args);
	}

	@Bean
	ApplicationRunner runner() {
		return args -> {
            kafkaTemplate.send(TEST_TOPIC, TEST_MESSAGE);
            logger.info("Message {} sent to topic {}", TEST_TOPIC, TEST_MESSAGE);
          };
	}

	Logger logger = LoggerFactory.getLogger(this.getClass());

	@KafkaListener(topics = TEST_TOPIC)
	public void consume(@Payload String message, @Header(KafkaHeaders.RECEIVED_TOPIC) String topic) {
		logger.info("Received message {} from topic {}", message, topic);
	}

	@Configuration
	static class DelayedConfiguration {

		@Bean
		public ConsumerFactory<Object, Object> consumerFactory() {
			Map<String, Object> config = new HashMap<>();
			config.put(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, "localhost:9092");
			config.put(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class);
			config.put(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class);
			config.put(ConsumerConfig.GROUP_ID_CONFIG, "grp_id");
			config.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest");
			return new DefaultKafkaConsumerFactory<>(config);
		}

		@Bean
		public ConcurrentKafkaListenerContainerFactory<Object, Object> kafkaListenerContainerFactory(ConsumerFactory<Object, Object> consumerFactory,
																									 ListenerContainerRegistry registry, TaskScheduler scheduler) {
			ConcurrentKafkaListenerContainerFactory<Object, Object> factory = new ConcurrentKafkaListenerContainerFactory<>();
			factory.setConsumerFactory(consumerFactory);
			KafkaConsumerBackoffManager backOffManager = createBackOffManager(registry, scheduler);
			factory.setContainerCustomizer(container -> {
				DelayedMessageListenerAdapter<Object, Object> delayedAdapter = wrapWithDelayedMessageListenerAdapter(backOffManager, container);
				delayedAdapter.setDelayForTopic(TEST_TOPIC, Duration.ofSeconds(6));
				delayedAdapter.setDefaultDelay(Duration.ofSeconds(1));
				container.setupMessageListener(delayedAdapter);
			});
			return factory;
		}

		@Bean
		public TaskScheduler taskScheduler() {
			return new ThreadPoolTaskScheduler();
		}

		@SuppressWarnings("unchecked")
		private DelayedMessageListenerAdapter<Object, Object> wrapWithDelayedMessageListenerAdapter(KafkaConsumerBackoffManager backOffManager, ConcurrentMessageListenerContainer<Object, Object> container) {
			return new DelayedMessageListenerAdapter<>((MessageListener<Object, Object>) container.getContainerProperties().getMessageListener(), backOffManager, container.getListenerId());
		}

		private ContainerPartitionPausingBackOffManager createBackOffManager(ListenerContainerRegistry registry, TaskScheduler scheduler) {
			return new ContainerPartitionPausingBackOffManager(registry, new ContainerPausingBackOffHandler(new ListenerContainerPauseService(registry, scheduler)));
		}

	}

}
```

Please let me know if this works for you and if you have any questions or suggestions.

Thanks!
